Refactor workInternal

wirew0rm opened the pull request:
This PR aims to restructure the core work function dispatch mechanism implemented in `Block::workInternal`.

Noteworthy changes:
- breaks up the code into determining available samples and afterwards dispatching to the correct work function
- restructure and partially reimplement decimate and stride functionalities:
  - the decimation settings (numerator/denominator) now determine the minimum atomic amount of work that can be processed (previously this was determined by the ports' minimum samples)
  - stride is applied after denominator samples. (Previously controlled by the ports' maxium samples)
- cleanup
  - reduce (gcc) warnings to 0

Open issues:
- [x] clang/emscirpten test failures
- [ ] treatment of incomplete last chunk
- [ ] support for moving tags to the first tag of this or the next tag if port constraints don't allow chunking.
- [x] remove debug printouts
- [x] squash git history
- [x] investigate and reenable asan

RalphSteinhagen commented on core/include/gnuradio-4.0/Port.hpp:
```diff
@@ -955,7 +956,7 @@ template<typename T>
 concept TagPredicate = requires(const T &t, const Tag &tag, Tag::signed_index_type readPosition) {
     { t(tag, readPosition) } -> std::convertible_to<bool>;
 };
-inline constexpr TagPredicate auto defaultTagMatcher    = [](const Tag &tag, Tag::signed_index_type readPosition) noexcept { return tag.index >= readPosition || tag.index < 0; };
```
Why the drop of `tag.index < 0`? This used to indicate that a tag should be applied immediately regardless of other index constraints.

wirew0rm replied:
This predicate was and is used to get the next matching tag and the matching tag after that in `nSamplesUntilNextTag(port, offset=0)`, where offset is the default of 0 for getting the first tag and 1 for getting the Tag not at position 1. But since `offset` ends up in `readPosition` of the predicate tag index only, as soon a there is  a tag with index < 0, it will also be return when looking for the second tag... I did not follow up more detailed how this worked previously, re-adding it makes qa_Block spin indefinitely.

Tag propagation and setting _cachedTag is not influenced by this since there the `-1` condition is checked separately in `Port::getTags(untilOffset)`.

I'm not completely sure on whether this is the correct change, this is just the rationale for removing it. I tried to keep the changes outside of the workInternal (and its subfunctions) small, as i've been bitten sometimes by doing extensive changes thoughout the codebase at too many places at once.

drslebedev commented on core/include/gnuradio-4.0/Block.hpp:
```diff
@@ -1292,6 +1137,223 @@ class Block : public lifecycle::StateMachine<Derived>, //
+            }
+        };
+        for_each_port([&adjust_for_input_port](PortLike auto &port) { adjust_for_input_port(port); }, std::forward<decltype(ports)>(ports));
+        return std::tuple{minSync, maxSync, maxAvailable, hasAsync};
```
Would it be better if one returns a struct instead of tuple?
```cpp
struct PortLimits {
   std::size_t minSync{};
   std::size_t maxSync{};
   std::size_t maxAvailable{};
   bool hasAsync{};
};
```
And later one can do:
```cpp
const auto in = getPortLimits(inputPorts<PortType::STREAM>(&self()));
const auto out = getPortLimits(outputPorts<PortType::STREAM>(&self()));
```

RalphSteinhagen commented on core/include/gnuradio-4.0/Block.hpp:
```diff
@@ -1292,6 +1137,223 @@ class Block : public lifecycle::StateMachine<Derived>, //
+        return std::numeric_limits<std::size_t>::max();
+    }
+
+    work::Status
+    invokeProcessBulk(auto &inputSpans, auto &writersTuple) {
+        // cannot use std::apply because it requires tuple_cat(inputSpans, writersTuple). The latter doesn't work because writersTuple isn't copyable.
+        return [&]<std::size_t... InIdx, std::size_t... OutIdx>(std::index_sequence<InIdx...>, std::index_sequence<OutIdx...>) {
+            return self().processBulk(std::get<InIdx>(inputSpans)..., std::get<OutIdx>(writersTuple)...);
+        }(std::make_index_sequence<std::tuple_size_v<std::remove_cvref_t<decltype(inputSpans)>>>(), std::make_index_sequence<std::tuple_size_v<std::remove_cvref_t<decltype(writersTuple)>>>());
+    }
```
Not (yet) fully tested but the below could convert the present `std::vector<std::span>` or `std::vector<"ConsumableSpan">` to `std::span<std::span..> or `std::span<"ConsumableSpan">`

```cpp
work::Status invokeProcessBulk(auto &inputSpans, auto &writersTuple) {
    auto convertIfNeeded = [](auto&& arg) -> decltype(auto) {
        using T = std::decay_t<decltype(arg)>;
        if constexpr (std::is_same_v<T, std::vector<typename T::value_type>>>) {
            // convert std::vector to std::span of [Consumable,Producable]Spans
            return std::span<typename T::value_type>{arg.data(), arg.size()};
        } else if (...) {  // similar for std::array<...>
        } else { // pass-through for other types
            return std::forward<decltype(arg)>(arg);
        }
    };

    return [&]<std::size_t... InIdx, std::size_t... OutIdx>(std::index_sequence<InIdx...>, std::index_sequence<OutIdx...>) {
        return self().processBulk(convertIfNeeded(std::get<InIdx>(inputSpans))..., convertIfNeeded(std::get<OutIdx>(writersTuple))...);
    }(std::make_index_sequence<std::tuple_size_v<std::remove_cvref_t<decltype(inputSpans)>>>(), std::make_index_sequence<std::tuple_size_v<std::remove_cvref_t<decltype(writersTuple)>>>());
}
```

@drslebedev your view?

RalphSteinhagen replied:
N.B. this way we do not expose if the dynamic input ports/spans are stored as vector or array and makes IMO a more maintainable and future-proof  API. Feedback is welcome.

drslebedev replied:
Using `std::span` for an array of ports as the signature for processBulk() looks good to me. 
`std::span<std::span..>` is better than `std::vector<std::span..>`.

However, I recommend postponing this to the next pull request, as it requires further modifications in the unit tests.

RalphSteinhagen commented on core/include/gnuradio-4.0/Block.hpp:
```diff
@@ -1292,6 +1137,223 @@ class Block : public lifecycle::StateMachine<Derived>, //
+        if (nOutSamplesBeforeRequestedStop > 0) {
+            return std::tuple{work::Status::OK, nOutSamplesBeforeRequestedStop, nOutSamplesBeforeRequestedStop};
+        }
+        return std::tuple{work::Status::OK, processed, processed};
```
return struct with named fields.

RalphSteinhagen commented on core/include/gnuradio-4.0/Block.hpp:
```diff
@@ -1312,7 +1374,35 @@ class Block : public lifecycle::StateMachine<Derived>, //
+     *   - settings
+     *      - apply and reset cached/merged tag
+     *   - get available samples count
+     *     - M: samples to consume on SYNC ports
+     *     - N: samples to produce on SYNC ports
```
need easier/more descriptive names than 'M' and 'N'

drslebedev commented on core/include/gnuradio-4.0/Block.hpp:
```diff
@@ -1292,6 +1137,223 @@ class Block : public lifecycle::StateMachine<Derived>, //
+        std::size_t maxSync               = std::numeric_limits<std::size_t>::max();
+        std::size_t maxAvailable          = std::numeric_limits<std::size_t>::max();
+        bool        hasAsync              = false;
+        auto        adjust_for_input_port = [this, &minSync, &maxSync, &maxAvailable, &hasAsync]<PortLike Port>(Port &port) {
```
```suggestion
        auto        adjustForInputPort = [this, &minSync, &maxSync, &maxAvailable, &hasAsync]<PortLike Port>(Port &port) {
```

RalphSteinhagen commented on core/include/gnuradio-4.0/Block.hpp:
```diff
@@ -1337,290 +1426,119 @@ class Block : public lifecycle::StateMachine<Derived>, //
-#endif
-
-            if (isEOSTagPresent || lifecycle::isShuttingDown(this->state())) {
+        // evaluate number of available and processable samples
+        const auto [minSyncIn, maxSyncIn, maxSyncAvailableIn, hasAsyncIn]     = getPortLimits(inputPorts<PortType::STREAM>(&self()));
+        const auto [minSyncOut, maxSyncOut, maxSyncAvailableOut, hasAsyncOut] = getPortLimits(outputPorts<PortType::STREAM>(&self()));
+        auto [hasTag, nextTag, nextEosTag, asyncEoS]                          = getTagPositions();
+        auto       maxChunk                                                   = getMergedBlockLimit(); // handle special cases for merged blocks. TODO: evaluate if/how we can get rid of these
+        const auto inputSkipBefore                                            = inputSamplesToSkipBeforeNextChunk(std::min({ maxSyncAvailableIn, nextTag, nextEosTag }));
+        const auto availableToProcess          = std::min({ maxSyncIn, maxChunk, (maxSyncAvailableIn - inputSkipBefore), (nextTag - inputSkipBefore), (nextEosTag - inputSkipBefore) });
+        const auto availableToPublish          = std::min({ maxSyncOut, maxSyncAvailableOut });
+        const auto [decimatedIn, decimatedOut] = decimate(minSyncIn, availableToProcess, minSyncOut, availableToPublish);
```
I like the structured binding which makes the code more readable. Maybe these names could be also used in the generating methods w.r.t. naming which make those also more readable.

drslebedev replied:
```suggestion
        const auto [resampledIn, resampledOut] = decimate(minSyncIn, availableToProcess, minSyncOut, availableToPublish);
```

drslebedev commented on core/include/gnuradio-4.0/Block.hpp:
```diff
@@ -1292,6 +1137,223 @@ class Block : public lifecycle::StateMachine<Derived>, //
+        std::size_t nextTag               = std::numeric_limits<std::size_t>::max();
+        std::size_t nextEosTag            = std::numeric_limits<std::size_t>::max();
+        bool        asyncEoS              = false;
+        auto        adjust_for_input_port = [this, &hasTag, &nextTag, &nextEosTag, &asyncEoS]<PortLike Port>(Port &port) {
```
```suggestion
        auto        adjustForInputPort = [this, &hasTag, &nextTag, &nextEosTag, &asyncEoS]<PortLike Port>(Port &port) {
```

drslebedev commented on core/include/gnuradio-4.0/Block.hpp:
```diff
@@ -1292,6 +1137,223 @@ class Block : public lifecycle::StateMachine<Derived>, //
+            }
+        };
+        for_each_port([&adjust_for_input_port](PortLike auto &port) { adjust_for_input_port(port); }, inputPorts<PortType::STREAM>(&self()));
+        return std::tuple{hasTag, nextTag, nextEosTag, asyncEoS};
```
Would it be better if one returns a struct instead of tuple? (see previous comment on getPortLimits())

drslebedev commented on core/include/gnuradio-4.0/Block.hpp:
```diff
@@ -1292,6 +1137,223 @@ class Block : public lifecycle::StateMachine<Derived>, //
+    }
+
+    auto
+    decimate(std::size_t minSyncIn, std::size_t maxSyncIn, std::size_t minSyncOut, std::size_t maxSyncOut) {
```
Considering that it can do both decimation and interpolation, wouldn't it be more appropriate to rename it to resample()? This also aligns with other names.

RalphSteinhagen reviewed (approved):
@wirew0rm kudos, this was a tremendous amount of work to get it done this neatly and cleanly. :+1: 

The changes make the code much more readable and the now linear flow of processing w/o the earlier jumps easier to follow, to test and -- I believe -- also to maintain.

I added some small comments and suggestions for improvements. These are minor and the PR could from my point-of-view be merged as is, but I think these are quality items that would be beneficial and/or at least should be discussed/documented for a follow-up.

Main things are:
 * use of structs and named fields for function returns rather than std::tuple. This would complement and mirror the already used structured bindings.
 * use of `emitErrorMessage(..)` rather than throwing exceptions in core-lib code
 
 The others things and missing features you mentioned in the PR description could be done in a later follow-up.

drslebedev commented on core/include/gnuradio-4.0/Block.hpp:
```diff
@@ -1292,6 +1137,223 @@ class Block : public lifecycle::StateMachine<Derived>, //
+                } else {
+                    nResamplingChunks = std::min(maxSyncIn / denominator, maxSyncOut / numerator);
+                }
+                if (nResamplingChunks * denominator < minSyncIn || nResamplingChunks * numerator < minSyncOut) {
```
We might need to incorporate additional conditions to handle additional ill-formed parameters related to `numerator` and `denominator`. For instance, consider the scenario where `minSyncIn = 2`, `maxSyncIn = 3`, and `denominator = 2`. In this case, `nResamplingChunks = 1`, leading to a situation where it consistently returns {0, 0}, potentially resulting in an infinite loop.

drslebedev commented on core/include/gnuradio-4.0/Block.hpp:
```diff
@@ -1292,6 +1137,223 @@ class Block : public lifecycle::StateMachine<Derived>, //
+    }
+
+    work::Status
+    invokeProcessOneSimd(auto &inputSpans, auto &writersTuple, auto width, std::size_t processed) {
```
```suggestion
    invokeProcessOneSimd(auto &inputSpans, auto &writersTuple, auto width, std::size_t nSamplesToProcess) {
```

drslebedev commented on core/include/gnuradio-4.0/Block.hpp:
```diff
@@ -1292,6 +1137,223 @@ class Block : public lifecycle::StateMachine<Derived>, //
+    }
+
+    work::Status
+    invokeProcessOnePure(auto &inputSpans, auto &writersTuple, std::size_t processed) {
```
```suggestion
    invokeProcessOnePure(auto &inputSpans, auto &writersTuple, std::size_t nSamplesToProcess) {
```

drslebedev commented on core/include/gnuradio-4.0/Block.hpp:
```diff
@@ -1292,6 +1137,223 @@ class Block : public lifecycle::StateMachine<Derived>, //
+    }
+
+    auto
+    invokeProcessOneNonConst(auto &inputSpans, auto &writersTuple, std::size_t processed) {
```
```suggestion
    invokeProcessOneNonConst(auto &inputSpans, auto &writersTuple, std::size_t nSamplesToProcess) {
```

drslebedev commented on core/include/gnuradio-4.0/Block.hpp:
```diff
@@ -1337,290 +1426,119 @@ class Block : public lifecycle::StateMachine<Derived>, //
+        const auto   inputSpans   = prepareStreams(inputPorts<PortType::STREAM>(&self()), limitByFirstTag ? 1 : decimatedIn);
+        auto         writersTuple = prepareStreams(outputPorts<PortType::STREAM>(&self()), limitByFirstTag ? 1 : decimatedOut);
+        work::Status ret;
+        std::size_t  processed    = limitByFirstTag ? 1 : decimatedIn;
```
```suggestion
        std::size_t  nSamplesToProcess    = limitByFirstTag ? 1 : decimatedIn;
```

drslebedev commented on core/include/gnuradio-4.0/Block.hpp:
```diff
@@ -1337,290 +1426,119 @@ class Block : public lifecycle::StateMachine<Derived>, //
+        auto         writersTuple = prepareStreams(outputPorts<PortType::STREAM>(&self()), limitByFirstTag ? 1 : decimatedOut);
+        work::Status ret;
+        std::size_t  processed    = limitByFirstTag ? 1 : decimatedIn;
+        std::size_t  processedOut = limitByFirstTag ? 1 : decimatedOut;
```
```suggestion
        std::size_t  nSamplesToProcessOut = limitByFirstTag ? 1 : decimatedOut;
```

wirew0rm replied:
These are the only 2 vairables where I didn't manage to make them const, otherwise i'd agree with the rename, but these are also updated with the result of the processing...
What i'm missing here is the ability to define an uninitialized const variable and then have the compiler check that it will be assigned exactly once in every branch... Java does that, but in c++ the only way to do get something like that is to wrap it in a lambda and return from there, but that would make the code a lot less straightforward to read...

drslebedev commented on core/test/qa_Block.cpp:
```diff
@@ -557,8 +556,8 @@ stride_test(const StrideTestData &data, std::shared_ptr<gr::thread_pool::BasicTh
-    auto sched = scheduler(std::move(flow), thread_pool);
-    sched.runAndWait();
+    auto sched = scheduler(std::move(flow), std::move(thread_pool));
+    expect(sched.runAndWait().has_value());
```
Not for this PR, but I think we need to add this `expect(sched.runAndWait().has_value());` for all test, potentially it saves debugging time as it was the case here.

drslebedev reviewed (approved):
Thanks for the PR! 
Really great work to simplify the code and make it clear and readable!
Since we already discussed your changes privately. I added some minor comments, please feel free to implement them.

frankosterfeld commented on core/include/gnuradio-4.0/Block.hpp:
```diff
@@ -1337,290 +1426,119 @@ class Block : public lifecycle::StateMachine<Derived>, //
-            const bool success = consumeReaders(self(), nSamplesToConsume);
-            return { requested_work, ports_status.in_samples, success ? ret : work::Status::ERROR };
-
+            ret = invokeProcessBulk(inputSpans, writersTuple); // todo: evaluate how many where really produced...
```
typo: where/were

wirew0rm commented on core/include/gnuradio-4.0/Block.hpp:
```diff
@@ -818,69 +723,85 @@ class Block : public lifecycle::StateMachine<Derived>, //
+
+        const auto [minSyncIn, maxSyncIn, _, _1]    = getPortLimits(inputPorts<PortType::STREAM>(&self()));
+        const auto [minSyncOut, maxSyncOut, _2, _3] = getPortLimits(outputPorts<PortType::STREAM>(&self()));
+        if (minSyncIn > maxSyncIn) {
+            throw std::invalid_argument(fmt::format("Min samples for input ports ({}) is larger then max samples for input ports ({})", minSyncIn, maxSyncIn));
+        }
+        if (minSyncOut > maxSyncOut) {
+            throw std::invalid_argument(fmt::format("Min samples for output ports ({}) is larger then max samples for output ports ({})", minSyncOut, maxSyncOut));
+        }
+        if (denominator > maxSyncIn) {
+            throw std::invalid_argument(fmt::format("resampling denominator ({}) is larger then max samples for input ports ({})", denominator, maxSyncIn));
+        }
+        if (numerator > maxSyncOut) {
+            throw std::invalid_argument(fmt::format("resampling numerator ({}) is larger then max samples for output ports ({})", numerator, maxSyncOut));
+        }
```
:heavy_check_mark: done

wirew0rm commented on core/include/gnuradio-4.0/Block.hpp:
```diff
@@ -1292,6 +1137,223 @@ class Block : public lifecycle::StateMachine<Derived>, //
+            }
+        };
+        for_each_port([&adjust_for_input_port](PortLike auto &port) { adjust_for_input_port(port); }, std::forward<decltype(ports)>(ports));
+        return std::tuple{minSync, maxSync, maxAvailable, hasAsync};
```
:ballot_box_with_check: done

wirew0rm commented on core/include/gnuradio-4.0/Block.hpp:
```diff
@@ -1292,6 +1137,223 @@ class Block : public lifecycle::StateMachine<Derived>, //
+                }
+            }
+        }
+        return inputSamplesToSkipBefore;
```
looks good. I changed from the lambda to a simple bool, since it has no arguments and it is the first thing to be eavauated in the if so I didn't see the benefit of not directly evaluating the boolean expression. Let me know if I overlooked something there.

wirew0rm commented on core/include/gnuradio-4.0/Block.hpp:
```diff
@@ -1292,6 +1137,223 @@ class Block : public lifecycle::StateMachine<Derived>, //
+            std::size_t n = std::min(maxSyncIn, maxSyncOut);
+            return std::tuple{n, n};
+        }
+    }
```
done! I like the early returns and reduced nesting.
I thought about simplifying this before. With the first constexpr if, I was not sure if I can mix compile time and runtime expressions inside of one statement (denominator/numerator may be compile time constants or run-time member fields depending as chosen by the block author).

wirew0rm commented on core/include/gnuradio-4.0/Block.hpp:
```diff
@@ -818,69 +723,85 @@ class Block : public lifecycle::StateMachine<Derived>, //
+                    [nSamples]<typename OutputRange>(auto, OutputRange &outputRange) {
+                        auto processOneRange = [nSamples]<typename Out>(Out &out) {
                             if constexpr (Out::isMultiThreadedStrategy()) {
                                 if (!out.isFullyPublished()) {
-                                    fmt::print(stderr, "Block::write_to_outputs - did not publish all samples for MultiThreadedStrategy\n");
+                                    fmt::print(stderr, "Block::publishWriters - did not publish all samples for MultiThreadedStrategy\n");
                                     std::abort();
                                 }
                             }
```
I know we had some discussion about this with @drslebedev I don't remember what what the outcome was there.

wirew0rm commented:
Addressed most of the review comments, open points:

- naming numerator/denominator
- moving publish logic into producable span destructor
- using span for dynamic ports

drslebedev commented:
> * naming numerator/denominator
> * moving publish logic into producable span destructor
> * using span for dynamic ports

At least second and third points should be addressed in the next PRs.

frankosterfeld commented on core/include/gnuradio-4.0/Block.hpp:
```diff
@@ -807,80 +712,116 @@ class Block : public lifecycle::StateMachine<Derived>, //
-                throw std::runtime_error(fmt::format("Block is not defined as `ResamplingRatio<>`, but numerator = {}, denominator = {}, they both must equal to 1.", numerator, denominator));
+                auto e = gr::Error(fmt::format("Block is not defined as `ResamplingRatio<>`, but numerator = {}, denominator = {}, they both must equal to 1.", numerator, denominator));
+                emitErrorMessage("Block::checkParametersAndThrowIfNeeded:", e);
+                requestStop();
```
Doesn't this return DONE where it should return ERROR?

frankosterfeld commented on core/include/gnuradio-4.0/Block.hpp:
```diff
@@ -1337,290 +1451,127 @@ class Block : public lifecycle::StateMachine<Derived>, //
+        // if the block state changed to DONE, publish EOS tag on the next sample
+        if (ret == work::Status::DONE) {
+            this->setAndNotifyState(lifecycle::State::STOPPED);
+            publishTag({ { gr::tag::END_OF_STREAM, true } }, 1);
```
Why index 1 here, and not 0 (like in other places?)?

wirew0rm replied:
The done tag always gets the index of the next sample (which will never be published). Otherwise a block could not publish EoS without publishing more samples. But you are right, that seems to be already considered in publishTag, it does not publish to the last written but to the next written index. Probably it doesn't matter because in the next iteration it sees the state and publishes another tag on the correct position... so maybe it should just be removed s.t. all eos tag handling is done at the beginning of the function...

frankosterfeld commented on core/include/gnuradio-4.0/Block.hpp:
```diff
@@ -1337,290 +1451,127 @@ class Block : public lifecycle::StateMachine<Derived>, //
+        if (ret == work::Status::DONE) {
+            this->setAndNotifyState(lifecycle::State::STOPPED);
+            publishTag({ { gr::tag::END_OF_STREAM, true } }, 1);
+            ret = work::Status::DONE;
```
This line should be necessary (ret already DONE)

wirew0rm replied:
there probably was a second condition in the if before... can be removed when fixing the eos behavior for process bulk that doesn't consume anything

frankosterfeld commented on core/include/gnuradio-4.0/Block.hpp:
```diff
@@ -1337,290 +1451,127 @@ class Block : public lifecycle::StateMachine<Derived>, //
+            static_assert(gr::meta::always_false<gr::traits::block::stream_input_port_types_tuple<Derived>>, "neither processBulk(...) nor processOne(...) implemented");
+        }
+        forwardTags();
+        if (lifecycle::isShuttingDown(this->state()) || nextEosTag <= processedIn + 1) {
```
Here something is going wrong for my Delay block (processBulk(ConsumableSpan, PublishableSpan): The block produces/consumes 0, but both `processedIn` and `processedOut` are `100`, which is the data available at the input port. This check then considers the EOS tag in range and finishes.

wirew0rm replied:
Yes, see also this comment: https://github.com/fair-acc/graph-prototype/pull/297/files#diff-d0a6abe85a332778390dc74e1c498e4c580677fe46db76395cdf2afefc41fb1eR1507 Until now, this has not been a problem, because there was no block that relied on being able to not consume all input samples in process bulk. Probably the easiest and most generic way to fix it would be to remove the eosTag condition and only shutdown the block in the next call to process bulk...  have to check if this breaks something else.
